### ISSUE TITLE
[BUG] Warnings/errors running LSTMFCNClassifier with attention=True

### ISSUE THREAD

**Issue by chrico-bu-uab** [sktime/sktime#6553]
I get the following warnings and errors. Sorry if hard to parse/read.

> WARNING:tensorflow:Gradients do not exist for variables ['attention_lstm/kernel:0', 'attention_lstm/recurrent_kernel:0', 'attention_lstm/attention_kernel:0', 'attention_lstm/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
> WARNING:tensorflow:Gradients do not exist for variables ['attention_lstm/kernel:0', 'attention_lstm/recurrent_kernel:0', 'attention_lstm/attention_kernel:0', 'attention_lstm/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?
> 2024-06-06 23:16:43.722963: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.
> 2024-06-06 23:16:44.830968: I tensorflow/core/framework/local_rendezvous.cc:421] Local rendezvous recv item cancelled. Key hash: 9497913403789017232
> [... more of these key hash warnings]

---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
Cell In[8], line 5
      1 import warnings
      3 warnings.filterwarnings("ignore")
----> 5 experiment = run_experiment(X, y, S, D, F)  

File ~/sktime/examples/severity_models.py:884, in run_experiment(X, y, S, D, F, **kwargs)
    883 def run_experiment(X, y, S, D, F, **kwargs):
--> 884     output = run_cv(X, y, S, D, F, **kwargs)
    885     with open("output.pkl", "wb") as f:
    886         pickle.dump(output, f)

File ~/sktime/examples/severity_models.py:878, in run_cv(X, y, S, D, F, n_splits, **kwargs)
    876 # Replace with multiprocessing if you'd like
    877 for rnd, (dev, val) in enumerate(indices):
--> 878     add_output(dev, val, X, y, S, D, F, n_splits, rnd, output, **kwargs)
    880 return output

File ~/sktime/examples/severity_models.py:853, in add_output(dev, val, X, y, S, D, F, n_splits, rnd, output, **kwargs)
    850 # Train the model and get the probabilities and SHAP values
    851 X_dev, X_val, y_dev, y_val, S_dev, S_val, D_dev, D_val, F_dev, F_val = X.loc[dev], X.loc[val], y.loc[dev], \
    852     y.loc[val], S.loc[dev], S.loc[val], D.loc[dev], D.loc[val], F.loc[dev], F.loc[val]
--> 853 probabilities, shap_values, model = run_model(X_dev, X_val, y_dev, y_val, S_dev, S_val, D_dev, D_val, F_dev, F_val,
    854                                               n_splits, rnd, **kwargs)
    855 output["probabilities"].append(probabilities)
    856 output["shap_values"].append(shap_values)

File ~/sktime/examples/severity_models.py:678, in run_model(X_dev, X_val, y_dev, y_val, S_dev, S_val, D_dev, D_val, F_dev, F_val, n_splits, rnd, **kwargs)
    676 def run_model(X_dev, X_val, y_dev, y_val, S_dev, S_val, D_dev, D_val, F_dev, F_val, n_splits, rnd, **kwargs):
    677     # augment S with predictions from time series models
--> 678     G_dev, G_val = get_base_preds(X_dev, y_dev, X_val, y_val, n_splits, rnd)
    679     S_dev = pd.concat([S_dev, G_dev], axis=1)
    680     S_val = pd.concat([S_val, G_val], axis=1)

File ~/sktime/examples/severity_models.py:668, in get_base_preds(X_dev, y_dev, X_val, y_val, n_splits, rnd)
    666     train = X_dev.index[train]
    667     test = X_dev.index[test]
--> 668     update_G(G_dev, (train, test, fold))
    670 # Update G_val using the full training set
    671 update_G(G_val)

File ~/sktime/examples/severity_models.py:630, in get_base_preds.<locals>.update_G(G, fold)
    628     test = X_val.index
    629 print(name, fold, X_train.shape, X_test.shape, X_test.iloc[0, 0].shape)
--> 630 clf.fit(X_train, y_train)
    631 # clf._is_fitted = True
    632 clf.model_ = load_model(get_filename(name, fold))

File ~/sktime/sktime/classification/base.py:261, in BaseClassifier.fit(self, X, y)
    256         raise AttributeError(
    257             "self.n_jobs must be set if capability:multithreading is True"
    258         )
    260 # pass coerced and checked data to inner _fit
--> 261 self._fit(X, y)
    262 self.fit_time_ = int(round(time.time() * 1000)) - start
    264 # this should happen last: fitted state is set to True

File ~/sktime/sktime/classification/deep_learning/lstmfcn.py:197, in LSTMFCNClassifier._fit(self, X, y)
    194 if self.verbose:
    195     self.model_.summary()
--> 197 self.history = self.model_.fit(
    198     X,
    199     y_onehot,
    200     batch_size=self.batch_size,
    201     epochs=self.n_epochs,
    202     verbose=self.verbose,
    203     callbacks=deepcopy(self._callbacks) if self._callbacks else None,
    204 )
    206 return self

File ~/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py:70, in filter_traceback.<locals>.error_handler(*args, **kwargs)
     67     filtered_tb = _process_traceback_frames(e.__traceback__)
     68     # To get the full stack trace, call:
     69     # `tf.debugging.disable_traceback_filtering()`
---> 70     raise e.with_traceback(filtered_tb) from None
     71 finally:
     72     del filtered_tb

File ~/sktime/venv/lib/python3.9/site-packages/tensorflow/python/eager/execute.py:60, in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     53   # Convert any objects of type core_types.Tensor to Tensor.
     54   inputs = [
     55       tensor_conversion_registry.convert(t)
     56       if isinstance(t, core_types.Tensor)
     57       else t
     58       for t in inputs
     59   ]
---> 60   tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
     61                                       inputs, attrs, num_outputs)
     62 except core._NotOkStatusException as e:
     63   if name is not None:

InvalidArgumentError: Graph execution error:

Detected at node model/attention_lstm/while/attention_lstm_cell/add defined at (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel_launcher.py", line 17, in <module>

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/traitlets/config/application.py", line 1053, in launch_instance

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelapp.py", line 737, in start

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/tornado/platform/asyncio.py", line 195, in start

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 596, in run_forever

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/asyncio/events.py", line 80, in _run

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 524, in dispatch_queue

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 513, in process_one

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 418, in dispatch_shell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 758, in execute_request

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/ipkernel.py", line 426, in do_execute

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/zmqshell.py", line 549, in run_cell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3046, in run_cell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3101, in _run_cell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3306, in run_cell_async

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3488, in run_ast_nodes

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3548, in run_code

  File "/var/folders/cw/4c395x2x5v196hz5867xdvfc0000gn/T/ipykernel_39573/2010087879.py", line 5, in <module>

  File "/Users/temp/sktime/examples/severity_models.py", line 884, in run_experiment

  File "/Users/temp/sktime/examples/severity_models.py", line 878, in run_cv

  File "/Users/temp/sktime/examples/severity_models.py", line 853, in add_output

  File "/Users/temp/sktime/examples/severity_models.py", line 678, in run_model

  File "/Users/temp/sktime/examples/severity_models.py", line 668, in get_base_preds

  File "/Users/temp/sktime/examples/severity_models.py", line 630, in update_G

  File "/Users/temp/sktime/sktime/classification/base.py", line 261, in fit

  File "/Users/temp/sktime/sktime/classification/deep_learning/lstmfcn.py", line 197, in _fit

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1783, in fit

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1377, in train_function

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1360, in step_function

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1349, in run_step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1126, in train_step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 589, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/base_layer.py", line 1149, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 96, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/functional.py", line 515, in call

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/functional.py", line 672, in _run_internal_graph

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/layers/rnn/base_rnn.py", line 556, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/base_layer.py", line 1149, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 96, in error_handler

  File "/Users/temp/sktime/sktime/networks/lstmfcn_layers.py", line 706, in call

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/layers/rnn/base_rnn.py", line 722, in call

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/backend.py", line 5170, in rnn

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/backend.py", line 5149, in _step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/layers/rnn/base_rnn.py", line 717, in step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/base_layer.py", line 1149, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 96, in error_handler

  File "/Users/temp/sktime/sktime/networks/lstmfcn_layers.py", line 397, in call

Detected at node model/attention_lstm/while/attention_lstm_cell/add defined at (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel_launcher.py", line 17, in <module>

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/traitlets/config/application.py", line 1053, in launch_instance

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelapp.py", line 737, in start

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/tornado/platform/asyncio.py", line 195, in start

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 596, in run_forever

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once

  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/asyncio/events.py", line 80, in _run

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 524, in dispatch_queue

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 513, in process_one

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 418, in dispatch_shell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/kernelbase.py", line 758, in execute_request

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/ipkernel.py", line 426, in do_execute

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/ipykernel/zmqshell.py", line 549, in run_cell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3046, in run_cell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3101, in _run_cell

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3306, in run_cell_async

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3488, in run_ast_nodes

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3548, in run_code

  File "/var/folders/cw/4c395x2x5v196hz5867xdvfc0000gn/T/ipykernel_39573/2010087879.py", line 5, in <module>

  File "/Users/temp/sktime/examples/severity_models.py", line 884, in run_experiment

  File "/Users/temp/sktime/examples/severity_models.py", line 878, in run_cv

  File "/Users/temp/sktime/examples/severity_models.py", line 853, in add_output

  File "/Users/temp/sktime/examples/severity_models.py", line 678, in run_model

  File "/Users/temp/sktime/examples/severity_models.py", line 668, in get_base_preds

  File "/Users/temp/sktime/examples/severity_models.py", line 630, in update_G

  File "/Users/temp/sktime/sktime/classification/base.py", line 261, in fit

  File "/Users/temp/sktime/sktime/classification/deep_learning/lstmfcn.py", line 197, in _fit

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1783, in fit

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1377, in train_function

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1360, in step_function

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1349, in run_step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 1126, in train_step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/training.py", line 589, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/base_layer.py", line 1149, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 96, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/functional.py", line 515, in call

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/functional.py", line 672, in _run_internal_graph

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/layers/rnn/base_rnn.py", line 556, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/base_layer.py", line 1149, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 96, in error_handler

  File "/Users/temp/sktime/sktime/networks/lstmfcn_layers.py", line 706, in call

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/layers/rnn/base_rnn.py", line 722, in call

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/backend.py", line 5170, in rnn

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/backend.py", line 5149, in _step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/layers/rnn/base_rnn.py", line 717, in step

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 65, in error_handler

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/engine/base_layer.py", line 1149, in __call__

  File "/Users/temp/sktime/venv/lib/python3.9/site-packages/keras/src/utils/traceback_utils.py", line 96, in error_handler

  File "/Users/temp/sktime/sktime/networks/lstmfcn_layers.py", line 397, in call

2 root error(s) found.
  (0) INVALID_ARGUMENT:  Incompatible shapes: [2,64,8] vs. [128,64,8]
	 [[{{node model/attention_lstm/while/attention_lstm_cell/add}}]]
	 [[Func/model/attention_lstm/while/body/_1/input/_1220/_54]]
  (1) INVALID_ARGUMENT:  Incompatible shapes: [2,64,8] vs. [128,64,8]
	 [[{{node model/attention_lstm/while/attention_lstm_cell/add}}]]
0 successful operations.
0 derived errors ignored. [Op:__inference_train_function_6679]

**Comment by fkiraly** [comment]
Would setting the `loss` explicitly fix this?

**Comment by chrico-bu-uab** [comment]
It’s already explicitly set:

def build_model(self, input_shape, n_classes, **kwargs):
    """
    Construct a compiled, un-trained, keras model that is ready for training.

    ----------
    input_shape : tuple
        The shape of the data fed into the input layer
    n_classes: int
        The number of classes, which shall become the size of the output
         layer
    Returns
    -------
    output : a compiled Keras Model
    """
    import tensorflow as tf
    from tensorflow import keras

    tf.random.set_seed(self.random_state)

    input_layers, output_layer = self._network.build_network(input_shape, **kwargs)

    output_layer = keras.layers.Dense(units=n_classes, activation="softmax")(
        output_layer
    )

    model = keras.models.Model(inputs=input_layers, outputs=output_layer)

    model.compile(
        loss="categorical_crossentropy",
        optimizer="adam",
        metrics=["accuracy"],
    )
    # .get_params() returns an empty list for callback.
    # inconsistent with function initial run where callbacks was set to None
    self._callbacks = self.callbacks or None

    return model

This is from lstmfcn.py.

On Jun 7, 2024, at 2:30 PM, Franz Király ***@***.***> wrote:

You don't often get email from ***@***.*** Learn why this is important<https://aka.ms/LearnAboutSenderIdentification>


Would setting the loss explicitly fix this?

**Comment by fkiraly** [comment]
```python

def build_model(self, input_shape, n_classes, **kwargs):
    """
    Construct a compiled, un-trained, keras model that is ready for training.
    ----------
    input_shape : tuple
        The shape of the data fed into the input layer
    n_classes: int
        The number of classes, which shall become the size of the output
         layer
    Returns
    -------
    output : a compiled Keras Model
    """
    import tensorflow as tf
    from tensorflow import keras

    tf.random.set_seed(self.random_state)

    input_layers, output_layer = self._network.build_network(input_shape, **kwargs)

    output_layer = keras.layers.Dense(units=n_classes, activation="softmax")(
        output_layer
    )

    model = keras.models.Model(inputs=input_layers, outputs=output_layer)

    model.compile(
        loss="categorical_crossentropy",
        optimizer="adam",
        metrics=["accuracy"],
    )
    # .get_params() returns an empty list for callback.
    # inconsistent with function initial run where callbacks was set to None
    self._callbacks = self.callbacks or None

    return model
```

**Comment by fkiraly** [comment]
That's strange then - any ideas, @sktime/core-developers?

**Comment by fkroeber** [comment]
Has there been any development on this since last year? I'm also encountering the issue

**Comment by fkiraly** [comment]
@Ankit-1204 also reported it here https://github.com/sktime/sktime/issues/7905 - we are having issues with reproducing the correct conditions, the error does occur on all systems, apparently.

Reports of your operating system and versions would be appreciated, @fkroeber!
